Tests/Rules/AvoidAssignmentToAutomaticVariable.tests.ps1

```diff
@@ -4,55 +4,72 @@ $ruleName = "PSAvoidAssignmentToAutomaticVariable"
+            @{ VariableName = 'ShellId'; ExpectedSeverity = 'Error' }
+            @{ VariableName = 'true'; ExpectedSeverity = 'Error' }
+            # Variables introuced only in PowerShell 6.0 have a Severity of Warning only
+            @{ VariableName = 'IsCoreCLR'; ExpectedSeverity = $excpectedSeverityForAutomaticVariablesInPowerShell6; OnlyPresentInCoreClr = $true }
+            @{ VariableName = 'IsLinux'; ExpectedSeverity = $excpectedSeverityForAutomaticVariablesInPowerShell6; OnlyPresentInCoreClr = $true }
+            @{ VariableName = 'IsMacOS'; ExpectedSeverity = $excpectedSeverityForAutomaticVariablesInPowerShell6; OnlyPresentInCoreClr = $true }
+            @{ VariableName = 'IsWindows'; ExpectedSeverity = $excpectedSeverityForAutomaticVariablesInPowerShell6; OnlyPresentInCoreClr = $true }
         )
 
-        It "Variable '<VariableName>' produces warning of severity error" -TestCases $testCases_ReadOnlyVariables {
-            param ($VariableName)
+        It "Variable <VariableName> produces warning of Severity <ExpectedSeverity>" -TestCases $testCases_ReadOnlyVariables {
+            param ($VariableName, $ExpectedSeverity)
 
             $warnings = Invoke-ScriptAnalyzer -ScriptDefinition "`$${VariableName} = 'foo'" | Where-Object { $_.RuleName -eq $ruleName }
             $warnings.Count | Should -Be 1
-            $warnings.Severity | Should -Be $readOnlyVariableSeverity
+            $warnings.Severity | Should -Be $ExpectedSeverity
         }
 
-        It "Using Variable '<VariableName>' as parameter name produces warning of severity error" -TestCases $testCases_ReadOnlyVariables {
-            param ($VariableName)
+        It "Using Variable <VariableName> as parameter name produces warning of Severity error" -TestCases $testCases_ReadOnlyVariables {
+            param ($VariableName, $ExpectedSeverity)
 
             [System.Array] $warnings = Invoke-ScriptAnalyzer -ScriptDefinition "function foo{Param(`$$VariableName)}" | Where-Object {$_.RuleName -eq $ruleName }
             $warnings.Count | Should -Be 1
-            $warnings.Severity | Should -Be $readOnlyVariableSeverity
+            $warnings.Severity | Should -Be $ExpectedSeverity
         }
 
-        It "Using Variable '<VariableName>' as parameter name in param block produces warning of severity error" -TestCases $testCases_ReadOnlyVariables {
-            param ($VariableName)
+        It "Using Variable <VariableName> as parameter name in param block produces warning of Severity error" -TestCases $testCases_ReadOnlyVariables {
+            param ($VariableName, $ExpectedSeverity)
 
             [System.Array] $warnings = Invoke-ScriptAnalyzer -ScriptDefinition "function foo(`$$VariableName){}" | Where-Object {$_.RuleName -eq $ruleName }
             $warnings.Count | Should -Be 1
-            $warnings.Severity | Should -Be $readOnlyVariableSeverity
+            $warnings.Severity | Should -Be $ExpectedSeverity
```

Comment:
I see the above pattern a lot, but it worries me. I would think that the `Where-Object` clause would obfuscate additional unexpected errors. Shouldn't we know the total amount of errors we expect and not have to filter for only what we expect?

Reply:
There is no 'right' answer, I think. There is the approach of having either isolated unit tests (with the benefit of lower maintenance) or doing integration testing as part of the test. We are still running all rules against it, so we would probably still get a test failure if one rule unexpectedly threw an exception but we would not have to adapt tests too much if rules changed. I will leave this decision up to you and would be happy to remove the `Where-Object`